brickschema/validate.py

```diff
@@ -0,0 +1,423 @@
+        return
+    # end of triplesForOneViolation()
+
+# end of class Validate()
+
+class ResultsSerialize():
+    """
+    Serializes violations with extra offender info.
+    """
+
+    def __init__(self, violationList, namespaceDict, output):
+        self.log = logging.getLogger()
+        self.log.setLevel(logging.DEBUG if hasattr(sys, '_called_from_test') else logging.WARNING)
+
+        self.violationList = violationList
+        self.outFile = output
+        self.namespaceDict = namespaceDict
+
+
+    # Serialize and streamline (remove @prefix lines) a grpah and append to output
+    def __appendGraph(self, msg, g):
+        if msg:
+            self.outFile.write(msg)
+        for n in self.namespaceDict:
+            g.bind(n, self.namespaceDict[n])
+
+        for b_line in g.serialize(format='ttl').splitlines():
+            line = b_line.decode('utf-8')
+            # skip prefix, offendingTriple and blank line
+            if (not line.startswith('@prefix')) and \
+               ('offenderHint' not in line) and \
+               ('offendingTriple' not in line) and \
+               line.strip():
+                self.outFile.write(line)
+                self.outFile.write('\n')
+
+
+    def __appendViolation(self, msg, g):
+        # first print the violation body
+        self.__appendGraph(msg, g)
+
+        # tease out the triples with offendingTriple as predicate
+        tripleGraphs = []
+        tripleType = None
+        for (s, p, o) in g:
+            if p == BSH['offendingTriple'] or p == BSH['offenderHint']:
+                tripleType = p
+                tripleG = Graph()
+                for (s1, p1, o1) in o:
+                    tripleG.add((s1, p1, o1))
+                tripleGraphs.append(tripleG)
+
+        if len(tripleGraphs) == 0:
+            self.outFile.write('Please let us know if the contraint violation information is insufficient.\n')
+            return
+
+        if tripleType == BSH['offenderHint']:
+            self.outFile.write('Violation hint (subject predicate cause):\n')
+        elif len(tripleGraphs) == 1:
+            self.outFile.write('Offending triple:\n')
+        else:
+            self.outFile.write('Potential offending triples:\n')
+        for tripleG in tripleGraphs:
+            self.__appendGraph(None, tripleG)
+
+
+    def appendToOutput(self):
+        """
+        Append the contraint violations with the extra info (offending triples
+        or offender hint) to the output file.
+        """
+        self.outFile.write('\nAdditional info (%d constraint violations with offending triples):\n' %
+                           len(self.violationList))
+
+        # Print each violation graph, find and print the offending triple(s), too
+        for g in self.violationList:
+            self.__appendViolation('\nConstraint violation:\n', g)
+
+# end of class ResultsSerialize()
+
+# __main to avoid being included in the api documentation
+def __main():
```

Comment:
Is this the `brick_validate` script logic? Or is that elsewhere? If this is the `brick_validate` script logic, it should go into a dedicated `bin/` directory in the repo

Reply:
I'll move it to bin.